physics/discrete_trajectory_body.hpp

```diff
@@ -178,10 +237,34 @@ void DiscreteTrajectory<Frame>::ForgetBefore(Instant const& time) {
   // Get an iterator denoting the first entry with time >= |time|.  Remove all
   // the entries that precede it.  This preserves any entry with time == |time|.
   auto it = timeline_.lower_bound(time);
+  if (downsampling_.has_value() &&
+      (it == timeline_.end() ||
+       downsampling_->start_of_dense_timeline->first < it->first)) {
+    // The start of the dense timeline will be invalidated.
+    downsampling_->start_of_dense_timeline = it;
+    downsampling_->dense_intervals =
+        std::distance(downsampling_->start_of_dense_timeline,
+                      timeline_.cend()) - 1;
```

Comment:
I would factor these 4 line in a function of `Downsampling` that would probably be usable in `Append` and in other places.

Reply:
Done.